src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -165,35 +166,46 @@ private BoundExpression CreateStringConcat(SyntaxNode originalSyntax, ArrayBuild
                 case (StringConcatenationRewriteKind.AllStringsOrChars, _):
                 case (StringConcatenationRewriteKind.InvolvesObjects, _):
 fallbackStrings:
-
-#pragma warning disable IDE0055// Fix formatting
-                    // All other cases can be handled here
-#pragma warning restore IDE0055// Fix formatting
-                    concatMember = visitedArguments.Count switch
-                    {
-                        2 => SpecialMember.System_String__ConcatStringString,
-                        3 => SpecialMember.System_String__ConcatStringStringString,
-                        4 => SpecialMember.System_String__ConcatStringStringStringString,
-                        >= 5 => SpecialMember.System_String__ConcatStringArray,
-                        _ => throw ExceptionUtilities.UnexpectedValue(visitedArguments.Count),
-                    };
-
                     for (int i = 0; i < visitedArguments.Count; i++)
                     {
                         visitedArguments[i] = ConvertConcatExprToString(visitedArguments[i]);
                     }
 
-                    var finalArguments = visitedArguments.ToImmutableAndFree();
+                    ImmutableArray<BoundExpression> finalArguments = visitedArguments.ToImmutableAndFree();
 
+                    MethodSymbol concatMethod;
                     if (finalArguments.Length > 4)
                     {
-                        var array = _factory.ArrayOrEmpty(_factory.SpecialType(SpecialType.System_String), finalArguments);
-                        finalArguments = [array];
+                        if (_compilation.Assembly.RuntimeSupportsInlineArrayTypes
+                            && TryGetSpecialTypeMethod(originalSyntax, SpecialMember.System_String__ConcatReadOnlySpanString, out concatMethod, isOptional: true))
+                        {
+                            finalArguments = [CreateAndPopulateSpanFromInlineArray(
```

Comment:
>CreateAndPopulateSpanFromInlineArray [](http://example.com/codeflow?start=46&length=36)

It looks like there is no guarantee that we are going to get an instance of ```InternalSpecialType.System_ReadOnlySpan_T``` back